akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

```diff
@@ -0,0 +1,642 @@
+      // Unregister is only used to keep the map growing too large, but otherwise it is not critical to process it
+      // timely. In fact, the only way the map could keep growing would mean that we dequeue Registers from the
+      // queue, but then we will eventually reach the Deregister message, too.
+      if (nextElem ne null) {
+        needWakeup = false
+        if (onEvent(nextElem)) tryProcessNext(firstAttempt = true)
+      } else {
+        needWakeup = true
+        // additional poll() to grab any elements that might missed the needWakeup
+        // and have been enqueued just after it
+        if (firstAttempt)
+          tryProcessNext(firstAttempt = false)
+      }
+    }
+
+    def isShuttingDown: Boolean = shuttingDown
+
+    // External API
+    def enqueue(ev: Event): Unit = {
+      queue.add(ev)
+      /*
+       * Simple volatile var is enough, there is no need for a CAS here. The first important thing to note
+       * that we don't care about double-wakeups. Since the "wakeup" is actually handled by an actor message
+       * (AsyncCallback) we don't need to handle this case, a double-wakeup will be idempotent (only wasting some cycles).
+       *
+       * The only case that we care about is a missed wakeup. The characteristics of a missed wakeup are the following:
+       *  (1) there is at least one message in the queue
+       *  (2) the consumer is not running right now
+       *  (3) no wakeupCallbacks are pending
+       *  (4) all producers exited this method
+       *
+       * From the above we can deduce that
+       *  (5) needWakeup = true at some point in time. This is implied by (1) and (2) and the
+       *      'tryProcessNext' method
+       *  (6) There must have been one producer that observed needWakeup = false. This follows from (4) and (3)
+       *      and the implementation of this method. In addition, this producer arrived after needWakeup = true,
+       *      since before that, every queued elements have been consumed.
+       *  (7) There have been at least one producer that observed needWakeup = true and enqueued an element and
+       *      a wakeup signal. This follows from (5) and (6), and the fact that either this method sets
+       *      needWakeup = false, or the 'tryProcessNext' method, i.e. a wakeup must happened since (5)
+       *  (8) If there were multiple producers satisfying (6) take the last one. Due to (6), (3) and (4) we know
+       *      there cannot be a wakeup pending, and we just enqueued an element, so (1) holds. Since we are the last
+       *      one, (2) must be true or there is no lost wakeup. However, due to (7) we know there was at least one
+       *      wakeup (otherwise needWakeup = true). Now, if the consumer is still running (2) is violated,
+       *      if not running then needWakeup = false is violated (which comes from (6)). No matter what,
+       *      contradiction. QED.
+       *
+       */
+      if (needWakeup) {
```

Comment:
Finally this has a proof. Not fully formal, but definitely proof in the usual sense.